repl/src/main/scala/ammonite/repl/Main.scala

```diff
@@ -70,21 +80,104 @@ case class Main(predef: String = "",
+            case Seq() =>
+              runScript(path, args, kwargs, storage, cacheTag)
+            case cachedData =>
+              val outBuffer = mutable.Buffer.empty[String]
+              val warningBuffer = mutable.Buffer.empty[String]
+              val errorBuffer = mutable.Buffer.empty[String]
+              val infoBuffer = mutable.Buffer.empty[String]
+              val prompt = Ref("@ ")
+              val colors = Ref[Colors](Colors.Default)
+              val frontEnd = Ref[FrontEnd](AmmoniteFrontEnd(Filter.empty))
+              val printer = Printer(
+                outBuffer.append(_),
+                warningBuffer.append(_),
+                errorBuffer.append(_),
+                infoBuffer.append(_)
+              )
+              val predef = Main.maybeDefaultPredef(defaultPredef, Main.defaultPredefString)
+
+              val interp: Interpreter = new Interpreter(
+                prompt,
+                frontEnd,
+                frontEnd().width,
+                frontEnd().height,
+                colors,
+                printer,
+                storage,
+                new History(Vector()),
+                predef,
+                wd,
+                Nil,
+                false
+              )
+              
+              def evalMain(cls: Class[_]) =
+                cls.getDeclaredMethod("$main").invoke(null)
+
+              var blockNumber = 1;
+              def getBlockNumber = blockNumber match {
+                case 1 => ""
+                case _ => "_" + blockNumber.toString
+              }
+
+              // Iterate through cached blocks and execute classFiles
+              // blockNumber keeps track of blockIndex
+              cachedData.foreach { d =>
+                for {
+                  cls <- interp.eval.loadClass(pkg + "." + wrapper + getBlockNumber, d._1)
+                } yield evalMain(cls)
+                blockNumber += 1
+              }
+
+              if (Timer.show)
+                println("Compilation Count: " + interp.compilationCount)
+          }
+        }
+        else {
+          runScript(path, args, kwargs, storage, cacheTag)
+        }
+    }
+  }
+
+
+
   /**
     * Run a Scala script file! takes the path to the file as well as an array
     * of `args` and a map of keyword `kwargs` to pass to that file.
     */
   def runScript(path: Path,
                 args: Seq[String],
-                kwargs: Map[String, String]): Res[Imports] = {
-
+                kwargs: Map[String, String],
+                storage: Storage,
+                cacheTag: String,
+                cacheScript: Boolean = true): Res[Seq[ImportData]] = {
     val repl = instantiateRepl()
     val (pkg, wrapper) = Util.pathToPackageWrapper(path, wd)
-    repl.interp.processModule(read(path), wrapper, pkg) match{
+    repl.interp.processModule(read(path), wrapper, pkg) match {
       case x: Res.Failing => x
-      case Res.Success(imports) =>
+      case Res.Success(data) =>
```

Comment:
data is list of cached classFiles and corresponding imports to load and eval classFiles